provider/block_devices_plug.py

```diff
@@ -0,0 +1,486 @@
+        self._orig_disks = set()
+        self._all_disks = set()
+        self._qmp_outputs = {}
+        self._event_devs = []
+        self._dev_type = DISK
+        self._qdev_type = qdevices.QBlockdevNode if vm.check_capability(
+            Flags.BLOCKDEV) else qdevices.QDrive
+        self._timeout = 300
+
+    def __getitem__(self, index):
+        """ Get the hot plugged disk index. """
+        return self._plugged_disks[index]
+
+    def __len__(self):
+        """ Get the len of the hot plugged disks. """
+        return len(self._plugged_disks)
+
+    def __iter__(self):
+        """ Iterate the the hot plugged disks. """
+        for disk in self._plugged_disks:
+            yield disk
+
+    def _list_all_disks(self):
+        """ List all the disks. """
+        session = self.vm.wait_for_login(timeout=360)
+        if self._islinux:
+            self._all_disks = utils_misc.list_linux_guest_disks(session)
+        else:
+            self._all_disks = set(session.cmd('wmic diskdrive get index').split()[1:])
+        session.close()
+        return self._all_disks
+
+    def _check_qmp_outputs(self, action):
+        """ Check the output of qmp commands. """
+        for dev_id in list(self._qmp_outputs.keys()):
+            output = self._qmp_outputs.pop(dev_id)
+            if output[1] is False:
+                raise TestError("Failed to %s device %s." % (action, dev_id))
+
+    def _get_events_deleted(self):
+        """ Get the device deleted events. """
+        self.event_devs = [img for img in self._unplugged_devs.keys()]
+        for event in self.vm.monitor.get_events():
+            if DELETED_EVENT in event.get("event") and 'device' in event.get('data'):
+                name = event.get('data')['device']
+                if name in self.event_devs:
+                    self.vm.monitor.clear_event(DELETED_EVENT)
+                    self.event_devs.remove(name)
+        return not self.event_devs
+
+    def _wait_events_deleted(self, timeout=300):
+        """
+        Wait the events "DEVICE DELETED" to be generated after unplug device.
+        """
+        if not utils_misc.wait_for(self._get_events_deleted, timeout):
+            raise TestError(
+                'No \"DEVICE DELETED\" event generated after unplug \"%s\".' %
+                (';'.join(self.event_devs)))
+
+    def _create_devices(self, images):
+        """ Create the block devcies. """
+        self._hotplugged_devs.clear()
+        for img in images:
+            bus_name = None
+            self._hotplugged_devs[img] = []
+            img_params = self.vm.params.object_params(img)
+            devices_created = getattr(
+                self.vm.devices, '%s_define_by_params' % self._dev_type['name'])(
+                img, img_params, self._dev_type['media'])
+
+            for dev in reversed(devices_created):
+                if dev.get_qid().endswith(img):
+                    self._hotplugged_devs[img].insert(-1, dev)
```

Comment:
LINE 236 and 242 are repeated, please put it outside the `if` clause.

Reply:
For line 242, It adds the hba device if needed.